A42-xds-ring-hash-lb-policy.md

```diff
@@ -0,0 +1,580 @@
+will produce the same hash.  Since a hash policy examines specific parts of
+a request, it can fail to produce a hash (e.g., if the hashed header is not
+present).  If (and only if) all configured hash policies fail to generate a
+hash, a random hash will be used for the request, which (assuming a
+hash-based LB policy is used) will result in picking a random endpoint
+for the request.  If a hash policy has the `terminal` attribute set to true,
+and the policy does generate a result, then all subsequent hash policies
+are skipped.
+
+These semantics allow gRPC to support only a subset of xDS hash policy
+types in a forward-compatible way.  For any hash policy type that gRPC
+does not support, that hash policy is treated as simply not returning any
+result for the request.
+
+Here is how gRPC will handle each type of hash policy:
+
+- [`header`](https://github.com/envoyproxy/envoy/blob/2443032526cf6e50d63d35770df9473dd0460fc0/api/envoy/config/route/v3/route_components.proto#L710):
+  gRPC will support this type of hash policy.  This allows hashing on a
+  request header.
+- [`cookie`](https://github.com/envoyproxy/envoy/blob/2443032526cf6e50d63d35770df9473dd0460fc0/api/envoy/config/route/v3/route_components.proto#L713):
+  gRPC does not support HTTP cookies, so we will not support this hash
+  policy.  If specified, this policy will not return any result.
+- [`connection_properties`](https://github.com/envoyproxy/envoy/blob/2443032526cf6e50d63d35770df9473dd0460fc0/api/envoy/config/route/v3/route_components.proto#L716):
+  In Envoy, this allows hashing based on the source IP address, which
+  ensures that all requests from a given client are sent to the same
+  endpoint.  However, gRPC cannot support this, because it is a client
+  and not a proxy, so it does not know what source address will be used
+  for the request until after it has finished load balancing.  If
+  specified, this policy will not return any result.  (However, see
+  `filter_state` below for an alternative.)
+- [`query_parameter`](https://github.com/envoyproxy/envoy/blob/2443032526cf6e50d63d35770df9473dd0460fc0/api/envoy/config/route/v3/route_components.proto#L719):
+  gRPC does not support query parameters, so we will not support this
+  hash policy.  If specified, this policy will not return any result.
+- [`filter_state`](https://github.com/envoyproxy/envoy/blob/2443032526cf6e50d63d35770df9473dd0460fc0/api/envoy/config/route/v3/route_components.proto#L722):
+  gRPC does not currently support the general-purpose concept of filter state
+  the way that Envoy does.  However, we will support one special filter state
+  [`key`](https://github.com/envoyproxy/envoy/blob/2443032526cf6e50d63d35770df9473dd0460fc0/api/envoy/config/route/v3/route_components.proto#L703)
+  called `io.grpc.channel_id`, which will hash to the same value for all
+  requests on a given gRPC channel.  This can be used in similar situations
```

Comment:
Call out that it is not 100% equivalent, since each channel will hash differently.

Reply:
Done.